### ISSUE TITLE
modular interface-based installation configurations

### ISSUE THREAD

**Issue by bendichter** [catalystneuro/neuroconv#94]
NeuroConv has many dependencies that are specific to particular data formats. We have been working on how to make imports modular. This works right now, but we might end up with a problem where different libraries have conflicting dependencies, so it would not be possible to install all of them. Therefore, I think it is also important to make installation modular. Currently, we have a few different installation options. You can run:

`pip install neuroconv`

for the minimal dependencies and 

`pip install neuroconv[full]`

for the full dependencies. I propose a solution where you can specify modalities during install and pip will know to install the dependencies specifically for those modalities, e.g.:

`pip install neuroconv[openephys]`

would install the minimum plus `pyopenephys`. A user would also be able to specify multiple dependencies, e.g.:

`pip install neuroconv[spikeglx,kilosort,phy,deeplabcut]`

and the environment would be set up with the necessary libraries for those data sources.

In order to accomplish this, we will need to formally define the requirements of each type of source data. One way we can do this is to have `requirements.txt` files in each data type directory. For instance: 
`neuroconv/src/neuroconv/datainterfaces/behavior/deeplabcut/requirements.txt` would be:

```
dlc2nwb>=0.2
```

The `setup.py` file would use `glob` to iterate through all of the datainterface modules and create a `extra_requires` `dict` for each source data type.

**Comment by CodyCBakerPhD** [comment]
Another approach could be to still only have one file at the top level (say, `extra_requires.json` or `extra_requires.pkl`) and what is saved within is a dictionary of format keywords and list of specific dependencies as values.

Main benefit would be readability as it would make it really easy to compare at a glance the dependencies across each formats - with the above approach I'd have to open separate tabs for each file on the web interface (on top of having to navigate down to each file), or similarly open each file separately (again, after navigating) on my local clone.

**Comment by bendichter** [comment]
I would rather not use pickle files. Not transparent to users and no version control. json is much better than pickle for this.

neo uses a dict in the setup.py file: https://github.com/NeuralEnsemble/python-neo/blob/50abfbce832edb1c0f8364ac34c8a6442e1a3edf/setup.py#L10-L18

I think I would still prefer the individual requirements.txt files. I know this is more work, but I'd like to go for more of a plugin style architecture where DataInterfaces can be contributed without altering any of the core files.

**Comment by bendichter** [comment]
Having separate `requirements.txt` files also allows for the option to install these options via `pip install -r path/to/requirements.txt

**Comment by h-mayorquin** [comment]
I really favor having a centralized place to list all the dependencies. I think that is easier for maintainers to get an overall sense of how the dependencies relate to each others easier to modify and test. It does not seem a big cost to me for developers to go and modify the centralized file -which is the only non sub-module file that should be modified- after they have done a change. Maybe I am missing something.

More importantly, I think that the centralized approach is more in line with where packing is going in the python community. My feeling is that since [PEP 621](https://peps.python.org/pep-0621/) the goal is to replace both `setup.py` and `requirements.txt` with a single specification in the toml file (this remove some problems that existed for build specification that you can read [here](https://snarky.ca/what-the-heck-is-pyproject-toml/)). **So, the community wants to move away from `setup.py` and `requierements.txt`**. The standard way of specifying extra-dependencies in the `toml` file is in this [link](https://setuptools.pypa.io/en/latest/userguide/pyproject_config.html) and you can see this is centralized. As an example of this trend the popular [poetry](https://python-poetry.org/) library uses the [toml file](https://python-poetry.org/docs/pyproject/) for package configuration and [adds the extras in the centralized fashion](https://stackoverflow.com/questions/60971502/python-poetry-how-to-install-optional-dependencies). 

What I see in #100 is that following the approach with multiple `requirements.txt` files is causing us to add further machinery to the `setup.py` but the latter is falling out of favor in the community. While the toml standard supports[ dynamically building the metadata ](https://setuptools.pypa.io/en/latest/userguide/pyproject_config.html#dynamic-metadata)(i.e. what we are doing with setup in #100 to add the extras during the build) it is not clear to me that we will able to export this approach cleanly or easy to toml if it does fulfill its role of becoming the standard and we want to migrate later.

If want still want to go the distributed approach as @bendichter is proposing I propose a gradual approach to this change. 
First we should migrate to the new standard toml approach with a centralized specification of extras and then, only after that, we should a the dynamically build specification with files distributed across the sub-packages. I also think this will be easier to review in PRs as first we will be introducing the testing for multiple installations and a more fine-grained specifications of the dependencies and then their dynamic construction in another PR. Right now #100 is doing the two at the same time which I think is too big of a change and you can see that in the diff.

**Comment by h-mayorquin** [comment]
> The first is that I envisioned this as sort of a plugin architecture where contributions of new modality DataInterfaces required no modifications to core or packaging code. This would make PRs more modular and it would enable us (and others) to create a DataInterface in a separate directory, importing base classes from NeuroConv, but never needing to make any modifications to the NeuroConv library at all.

Thank you, I understand what you mean by plug-in now. They can develop things independently and plug-back seamlessly if they want to. With the current approach they will copy the data interface to the appropriate place **and** have to modify any dependency that they added.

> The second reason is that it enables running `pip install -r path/to/requirements.txt` on any DataInterface requirements file. I'm not sure if this feature would ever be needed though since you could always just run `pip install .[modality]`.

Yep, I think we offer the same functionality with the  `pip install .[modality]`.

**Comment by CodyCBakerPhD** [comment]
I do love the `pyproject.toml` and it's nice to see more people advocating it. I'd be more than happy to swap other parts of the `setup` into there!

**Comment by h-mayorquin** [comment]
Another example in a plugin library:
https://github.com/imageio/imageio/blob/master/setup.py

**Comment by CodyCBakerPhD** [comment]
I do believe this was closed by #98 and others from around that time
